### PR TITLE
Fix/cdap 11044 hive table name

### PR DESCRIPTION
**Pull request by sheivin**
JIRA: https://issues.cask.co/browse/CDAP-11044

### CODE REVIEW

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java [hunk 1]
```diff
@@ -88,12 +89,11 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
   @Override
   protected Map<String, String> getProperties() {
-    Map<String, String> properties;
-    properties = new HashMap<>(tableSinkConfig.getProperties().getProperties());
-
-    properties.put(Properties.BatchReadableWritable.NAME, tableSinkConfig.getName());
-    properties.put(Properties.BatchReadableWritable.TYPE, Table.class.getName());
-    return properties;
+    ExploreProperties.Builder builder = ExploreProperties.builder();
+    builder.addAll(new HashMap<>(tableSinkConfig.getProperties().getProperties()));
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java]: you don't need to do new HashMap<>() here

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java [hunk 2]
```diff
@@ -88,12 +89,11 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
+    builder.setExploreTableName(tableSinkConfig.getName());
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java]: we should be adding a new optional property to the table source and sink called 'exploreName'. It should default to the name of the Table, but users should be able to set it to something else if they want. You do this by adding a new variable to the table source and sink configs and annotating it as nullable.

**Review by albertshau — COMMENTED** [review]
also, please add a unit test. You can look at ETLTPFSTestRun's testOrc() method for an example of how to deploy a pipeline in unit test and use the query client to run an explore query.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java (hunk 2 above)
Comment (sheivin) [reply core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java]: Understood. Will work on it :)

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java (hunk 1 above)
Comment (sheivin) [reply core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java]: Will do the necessary fixes. Thanks

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java [hunk 3]
```diff
@@ -88,12 +88,16 @@ public void initialize(BatchRuntimeContext context) throws Exception {
 
   @Override
   protected Map<String, String> getProperties() {
-    Map<String, String> properties;
-    properties = new HashMap<>(tableSinkConfig.getProperties().getProperties());
+    ExploreProperties.Builder builder = ExploreProperties.builder();
+    builder.addAll(tableSinkConfig.getProperties().getProperties());
+    if ((tableSinkConfig.exploreName != null) && (tableSinkConfig.exploreName.length() != 0)) {
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/TableSink.java]: `&& !tableSinkConfig.exploreName.isEmpty()`

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSinkConfig.java [hunk 4]
```diff
@@ -39,6 +39,12 @@
   @Description("The name of the record field that should be used as the row key when writing to the table.")
   private String rowField;
 
+  @Name(Properties.Table.EXPLORE_NAME)
+  @Description("Defaults to the name of the table but allows the user to set it to something else if they want.")
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSinkConfig.java]: The name of the explore table. Defaults to the name of the dataset.

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java [hunk 5]
```diff
@@ -235,14 +235,14 @@ public void testTableSinkWithMacro() throws Exception {
                                          ImmutableMap.of(Properties.Table.NAME, "${name}",
                                                          Properties.Table.PROPERTY_SCHEMA_ROW_FIELD, "rowkey"),
                                          null);
-    ETLStage sink = new ETLStage("tableSinkUnique", sinkConfig);
+    ETLStage sink = new ETLStage("tableSink", sinkConfig);
```
Comment (albertshau) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java]: any reason this test case is being changed?

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java [hunk 6]
```diff
@@ -265,4 +265,74 @@ public void testTableSinkWithMacro() throws Exception {
...
+    workflowManager.waitForRuns(ProgramRunStatus.COMPLETED, 1, 5, TimeUnit.MINUTES);
+
+    Connection connection = getQueryClient();
+
+    ResultSet resultsSource = connection.prepareStatement("select * from exploreTableSource").executeQuery();
+    resultsSource.next();
+    Assert.assertEquals("row1", resultsSource.getString(1));
+    Assert.assertEquals("samuel", resultsSource.getString(2));
+    Assert.assertEquals(5, resultsSource.getInt(3));
+
```
Comment (albertshau) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java]: assert that there are no more results here

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java [hunk 7]
```diff
@@ -265,4 +265,74 @@ public void testTableSinkWithMacro() throws Exception {
...
+    ResultSet resultsSink = connection.prepareStatement("select * from exploreTableSink").executeQuery();
+    resultsSink.next();
+    Assert.assertEquals("row1", resultsSink.getString(1));
+    Assert.assertEquals("samuel", resultsSink.getString(2));
+    Assert.assertEquals(5, resultsSink.getInt(3));
+
```
Comment (albertshau) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/sink/TableSinkTest.java]: assert that there are no more results here

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java [hunk 8]
```diff
@@ -127,6 +127,7 @@
    */
   public static class Table {
     public static final String NAME = "name";
+    public static final String EXPLORE_NAME = "exploreName";
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java]: the other properties in this class use the actual dataset properties, so we should be consistent and use ExploreProperties.PROPERTY_EXPLORE_TABLE_NAME here

**Review by albertshau — COMMENTED** [review]
minor code comments, looking good.

In addition to the code changes, you will find docs at core-plugins/docs/Table-batchsource.md and core-plugins/docs/Table-batchsink.md. You will also find a UI widgets file at core-plugins/widgets/Table-batchsource.json and core-plugins/widgets/Table-batchsink.json. Let's also update these files with the new property

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSinkConfig.java [hunk 9]
```diff
@@ -40,7 +40,7 @@
   private String rowField;
 
   @Name(Properties.Table.EXPLORE_NAME)
-  @Description("Defaults to the name of the table but allows the user to set it to something else if they want.")
+  @Description("Tha name of the explore table. It defaults to the name of the dataset but allows the user to set it to something else.")
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSinkConfig.java]: The

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSourceConfig.java [hunk 10]
```diff
@@ -45,7 +45,7 @@
   private String rowField;
 
   @Name(Properties.Table.EXPLORE_NAME)
-  @Description("Defaults to the name of the table but allows the user to set it to something else if they want.")
+  @Description("Tha name of the explore table. It defaults to the name of the dataset but allows the user to set it to something else.")
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/TableSourceConfig.java]: The

File: core-plugins/widgets/Table-batchsource.json [hunk 11]
```diff
@@ -13,8 +13,16 @@
...
           "widget-type": "textbox",
+          "label": "Explore Name",
+          "name": "explore.table.name"
+        },
+        {
+          "widget-type": "textbox",
           "label": "Row Field",
           "name": "schema.row.field"
+        },
+        {
```
Comment (albertshau) [inline core-plugins/widgets/Table-batchsource.json]: remove empty object

**Comment by sheivin** [comment]
Can't be merged till UI bug is fixed. (Reference: https://issues.cask.co/browse/CDAP-11556)

**Comment by sreevatsanraman** [comment]
Closing the PR since we won't hve the UI change in 4.2

**Pull request closed without merging**